File: 6 kyu/Help the bookseller !/solution.fs
let stockSummary(lstOfArt: string[]) (lstOf1stLetter: string[]): string =
    let format (x,y) =
        "(" + string(x) + " : " + string(y) + ")"
    
    if lstOfArt.Length = 0 || lstOf1stLetter.Length = 0 then 
        ""
    else
        let getBooks c = 
            lstOfArt 
                |> Array.filter(fun x -> x.StartsWith c)
                |> Array.map(fun x -> x.Split(' ').[1] |> int)
                |> Array.sum
        lstOf1stLetter
            |> Array.map (fun x -> format(x, getBooks x))
            |> String.concat " - "
________________________________________
let stockSummary(lstOfArt: string[]) (lstOf1stLetter: string[]): string =
    let (|Art|) (item : string) =
        match item.Split(' ') with
        | [| category; quantity |] -> string category.[0], int quantity
        | _ -> failwith "bad input"
    let art = lstOfArt |> Seq.map (|Art|) |> Seq.cache
    if art |> Seq.isEmpty |> not then
        let sumQuantity code =
            sprintf "(%s : %d)" code (Seq.sumBy (fun (category, quantity) -> if category = code then quantity else 0) art)
        lstOf1stLetter |> Seq.map sumQuantity |> String.concat " - "
    else
        ""
________________________________________
let artToStock (art: string []) (category: string) : int =
    art
    |> Array.filter (fun item -> item.StartsWith(category))
    |> Array.sumBy (fun a -> a.Split(' ').[1] |> int)

let stockSummary (art: string []) (categories: string []) : string =
    match art.Length with
    | 0 -> ""
    | _ ->
        categories
        |> Array.map
            (fun category ->
                "("
                + category
                + " : "
                + (artToStock art category).ToString()
                + ")")
        |> String.concat " - "
________________________________________
open System

let parseStockItem (stockItem : string) =
    let category = string <| stockItem.[0]
    let index = stockItem.IndexOf(' ')
    let quantity = int <| stockItem.Substring(index + 1)
    category, quantity

let formatSummary summary =
    summary
    |> Seq.map (fun (k, v) -> sprintf "(%s : %d)" k v)
    |> String.concat " - "

let stockSummary' stock categories =
    let stockByCategory =
        stock
        |> Seq.map parseStockItem
        |> Seq.groupBy fst
        |> Seq.map (fun (k, vs) -> k, vs |> Seq.sumBy snd)
        |> Map.ofSeq

    let getCategoryItem category =
        let quantity =
            Map.tryFind category stockByCategory
            |> Option.defaultValue 0
        category, quantity

    categories
    |> Seq.map getCategoryItem
    |> formatSummary

let stockSummary stock categories =
    match stock, categories with
    | [||], _
    | _ , [||] -> String.Empty
    | _ -> stockSummary' stock categories

File: 6 kyu/Help the bookseller !/solution.java
import java.util.Arrays;
import java.util.IntSummaryStatistics;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class StockList {

    private static class Book {
        public final String category;
        public final String code;
        public final int quantity;

        public Book(String label) {
            category = label.substring(0,1);
            code = label.split(" ")[0].substring(1);
            quantity = Integer.parseInt(label.split(" ")[1]);
        }
    }

    public static String stockSummary(String[] lstOfArt, String[] lstOf1stLetter) {
        if (lstOfArt.length == 0 || lstOf1stLetter.length == 0)
          return "";
        Map<String, Integer> categoryCounts = Arrays.stream(lstOfArt)
                .map(Book::new)
                .collect(Collectors.groupingBy(book -> book.category,
                         Collectors.summingInt(book -> book.quantity)));
        return Arrays.stream(lstOf1stLetter)
                .map(initial -> String.format("(%s : %d)", 
                                              initial, categoryCounts.getOrDefault(initial, 0)))
                .collect(Collectors.joining(" - "));
    }
}
________________________________________
import java.util.Map;
import static java.util.Arrays.stream;
import static java.util.stream.Collectors.groupingBy;
import static java.util.stream.Collectors.summingInt;
import static java.util.stream.Collectors.joining;

public class StockList {
  private static int stockCount(final String s) {
    return Integer.valueOf(s.split(" ")[1]);
  }
  
  public static String stockSummary(final String[] stock, final String[] categories) {
    if (stock.length == 0 || categories.length ==  0)
      return "";
    final Map<String, Integer> counts = stream(stock)
        .collect(groupingBy(s -> s.substring(0, 1), summingInt(StockList::stockCount)));
    return stream(categories)
        .map(s -> "(" + s + " : " + counts.getOrDefault(s, 0) + ")")
        .collect(joining(" - "));
  }
}
________________________________________
import java.util.stream.Stream;
import java.util.stream.Collectors;
public class StockList { 
  public static String stockSummary(String[] arts, String[] cats) {
    if (arts.length == 0) return "";
    final int space = arts[0].indexOf(" ");
    return Stream.of(cats)
      .map(c -> c + " : " + Stream.of(arts)
        .filter(a -> c.contentEquals(a.subSequence(0, 1)))
        .map(a -> a.substring(space + 1))
        .mapToInt(Integer::parseInt)
        .sum())
      .map(s -> "(" + s + ")")
      .collect(Collectors.joining(" - "));
  }
}
________________________________________
public class StockList { 
  public static String stockSummary(String[] lstOfArt, String[] lstOf1stLetter) {
    if (lstOfArt.length == 0 || lstOf1stLetter.length == 0) return "";
    
    int sum = 0;
    String res = "";
    
    for (String i : lstOf1stLetter) {
        sum = 0;
        for (String j : lstOfArt) {
            sum += j.substring(0,1).equals(i) ? Integer.parseInt(j.replaceAll("[^0-9]","")) : 0;
        }
        res += " - (" + i + " : " + sum + ")";
    }
    
    return res.substring(3);
  }
}
________________________________________
import static java.util.Arrays.stream;
import static java.util.stream.Collectors.groupingBy;
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.summingInt;

class StockList {
  static String stockSummary(String[] lstOfArt, String[] lstOf1stLetter) {
    if (lstOfArt.length > 0 && lstOf1stLetter.length > 0) {
      var counts = stream(lstOfArt).collect(groupingBy(s -> s.substring(0, 1), summingInt(s -> Integer.parseInt(s.split(" ")[1]))));
      return stream(lstOf1stLetter).map(s -> "(" + s + " : " + counts.getOrDefault(s, 0) + ")").collect(joining(" - "));
    }
    return "";
  }
}
